_posts/2020-08-31-rico-logging.md

```diff
@@ -0,0 +1,81 @@
+Have a look at both and maybe even a third solution.
+
+Simply dumping all the log messages into the central storage will give you a big pile of strings but not much information.
+In a centralized logging facility the context of the log becomes as important as the message.
+It is not sufficient to see a message "Disk almost full".
+You also need to know where this message originated, which software generated the message, what version was running at that time, which user was logged in and much more.
+
+This context needs to be collected on the system where the log message is created.
+So it is the responsibility of the application to provide this information together with the log message.
+It is a tedious work to collect and add this to each and every log.
+If this is done manually then most likely it will be forgotten more than once.
+
+Exactly here will Rico assist you and take all the burden from your shoulders.
+Rico provides the `RicoApplicationContext` which is basically a dictionary to store context information.
+Basic information such as application name, application version and much more are prefilled into the dictionary by Rico.
+Application specific information can be added to this dictionary as needed during runtime.  
+Such entries can be either global or just for the current thread.
+
+Now that all this context information is collected in the `RicoApplicationContext` all that is left to do is to include it in the log message.
+For this task Rico provides log appender for [Logback](http://logback.qos.ch/) and [Log4j2](https://logging.apache.org/log4j/2.x/).
+These appender take the message and place the entire content of the `RicoApplicationContext` to the MDC (Mapped Diagnostic Context).
+The MDC is also dictionary and meant to carry the context information of a log message.
+Now everything is ready to be sent to the centralized logging facility.
+
+![component diagram](/assets/posts/2020-08-31-rico-logging/logging-and-context.png)
+
+For sending log messages to a centralized logging facilities there exist many good solutions.
+One possibility is to use a log appender and send the message to the server instead of writing it to a file.
+Such appender are available for both Logback and Log4j.
+We have made good experience with both [logstash-gelf](https://github.com/mp911de/logstash-gelf) and [Logback GELF](https://github.com/osiegmar/logback-gelf).
+They use [GELF (Graylog Extended Log Format)](https://docs.graylog.org/en/latest/pages/gelf.html), a defacto standard to send log messages over the wire. 
+
+After all this theory we would like to also show you how this can look in action.
+A ready to run example is available in the [rico samples project](https://github.com/rico-projects/rico-samples/tree/master/logging-sample).
+It contains a Readme with all the information required to run it.
+There are also preconfigured docker images for both Graylog and ELK as well as sample applications using Logback and Log4j2.
```

Comment:
We should try to use the same color schema in the diagrams of the post. At the moment the boxes in the 2 diagrams use complete different colors (and even saturations)